src/bert_model.rs

```diff
@@ -0,0 +1,250 @@
+// Copyright 2018 The Google AI Language Team Authors and The HuggingFace Inc. team.
+// Copyright (c) 2018, NVIDIA CORPORATION.  All rights reserved.
+// Copyright (c) 2019 The sticker developers.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::borrow::Borrow;
+
+use failure::Fallible;
+use hdf5::{Dataset, File};
+use tch::nn::{self, Module, ModuleT, Path};
+use tch::Tensor;
+
+use crate::cow::CowTensor;
+use crate::hdf5_model::LoadFromHDF5;
+use crate::layers::{Dropout, Embedding, LayerNorm};
+
+/// Bert model configuration.
+#[derive(Debug)]
+pub struct BertConfig {
+    pub hidden_dropout_prob: f64,
+    pub hidden_size: i64,
+    pub layer_norm_eps: f64,
+    pub max_position_embeddings: i64,
+    pub type_vocab_size: i64,
+    pub vocab_size: i64,
+}
+
+/// Construct the embeddings from word, position and token_type embeddings.
+#[derive(Debug)]
+pub struct BertEmbeddings {
+    position_embeddings: Embedding,
+    token_type_embeddings: Embedding,
+    word_embeddings: Embedding,
+
+    layer_norm: LayerNorm,
+    dropout: Dropout,
+}
+
+impl BertEmbeddings {
+    /// Construct new Bert embeddings with the given variable store
+    /// and Bert configuration.
+    pub fn new<'a>(vs: impl Borrow<nn::Path<'a>>, config: &BertConfig) -> Self {
+        let vs = vs.borrow().sub("embeddings");
+
+        let word_embeddings = Embedding::new(
+            &vs,
+            "word_embeddings",
+            config.vocab_size,
+            config.hidden_size,
+        );
+
+        let position_embeddings = Embedding::new(
+            &vs,
+            "position_embeddings",
+            config.max_position_embeddings,
+            config.hidden_size,
+        );
+
+        let token_type_embeddings = Embedding::new(
+            &vs,
+            "token_type_embeddings",
+            config.type_vocab_size,
+            config.hidden_size,
+        );
+
+        let layer_norm = LayerNorm::new(
+            vs.sub("layer_norm"),
+            vec![config.hidden_size],
+            config.layer_norm_eps,
+            true,
+        );
+        let dropout = Dropout::new(config.hidden_dropout_prob);
+
+        BertEmbeddings {
+            word_embeddings,
+            position_embeddings,
+            token_type_embeddings,
+
+            layer_norm,
+            dropout,
+        }
+    }
+
+    pub fn forward(
```

Comment:
Why is BertEmbeddings not implementing `forward` via `Module`?

Reply:
Does `Module::forward` only support single `Tensor`?

Reply:
Indeed. So in those cases I just follow the naming conventions (so, it should probably be `forward_t` here).

Reply:
Renamed to `forward_t`.